src/autopas/containers/linkedCells/traversals/C04HCP.h

```diff
@@ -0,0 +1,198 @@
+void C04HCP<ParticleCell, PairwiseFunctor, dataLayout, useNewton3>::traverseParticlePairs() {
+  auto &cells = *(this->_cells);
+#if defined(AUTOPAS_OPENMP)
+#pragma omp parallel
+#endif
+  {
+    for (int color = 0; color < 4; ++color) {
+      traverseSingleColor(cells, color);
+
+#if defined(AUTOPAS_OPENMP)
+      if (color < 3) {
+#pragma omp barrier
+      }
+#endif
+    }
+  }  // close parallel region
+}
+
+/**
+ * Go through one color and search for cuboids belonging to the specified color.
+ * Uses shifts to go through the different dimensions and prevent overlapping of the cuboids.
+ *
+ * @tparam ParticleCell
+ * @tparam PairwiseFunctor
+ * @tparam dataLayout
+ * @tparam useNewton3
+ * @param cells
+ * @param color
+ */
+template <class ParticleCell, class PairwiseFunctor, DataLayoutOption::Value dataLayout, bool useNewton3>
+void C04HCP<ParticleCell, PairwiseFunctor, dataLayout, useNewton3>::traverseSingleColor(
+    std::vector<ParticleCell> &cells, int color) {
+  // determine a starting point of one of the grids
+  std::array<long, 3> startOfThisColor{};  // coordinates: {x,y,z}
+
+  // different starting points for different colors
+  // some colors are starting outside the grid because only part of their cuboids are part of the grid
+  // this way the starting points of sticking out cuboids can be determined as well
+  switch (color) {
+    case 0:
+      startOfThisColor = {0l, 0l, 0l};
+      break;
+    case 1:
+      startOfThisColor = {-4l, 0l, 1l};
+      break;
+    case 2:
+      startOfThisColor = {-4l, 0l, -2l};
+      break;
+    case 3:
+      startOfThisColor = {-2l, 0l, -1l};
+      break;
+  }
+
+  // to fix compiler complaints about perfectly nested loop.
+  long startX = startOfThisColor[0], endX = _end[0];
+  long startY = startOfThisColor[1], endY = _end[1];
+  long startZ = startOfThisColor[2], endZ = _end[2];
+
+  // iterate over cartesian grid
+#if defined(AUTOPAS_OPENMP)
+#pragma omp for schedule(dynamic, 1) collapse(3) nowait
+#endif
+  for (long z = startZ; z < endZ; z += 4) {
+    for (long y = startY; y < endY; y++) {
+      for (long x = startX; x < (endX + 4);
+           x += 6) {  // color starts every 6th column again, the +4 is needed to prevent ending too early, since it
+                      // will be shifted back inside the loop
+        long x_index = x;
+        // shift on x-axis according to z-value
+        // first: no shift
+        // second: -4 shift
+        // third: -2 shift
+        // fourth: go back to first
+        switch ((z - startZ) % 12 / 4) {
+          case 0:
+            break;
+          case 1:
+            x_index -= 4;
+            break;
+          case 2:
+            x_index -= 2;
+            break;
+        }
+        if ((y - startY) % 2 != 0) {  // shift x-axis every second y-row
+          x_index += 3;
+        }
+        const std::array<long, 3> base3DIndex = {x_index, y, z};
+        processBasePack6(cells, base3DIndex);
```

Comment:
```suggestion
        processBasePack6(cells, {x_index, y, z});
```
no need for temporary